File: 401_499/416_Partition_Equal_Subset_Sum/Solution.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum = 0, n = nums.size();
        for (int num : nums) sum += num;
        if (sum % 2 != 0) return false;
        sum = sum / 2;
        vector<bool> dp(sum + 1, false);
        dp[0] = true;

        for (int i = 0; i < n; i++) 
            for (int j = sum; j >= 0; j--) 
                if (j - nums[i] >= 0) 
                    dp[j] = dp[j] || dp[j - nums[i]];

        return dp[sum];
    }
};

File: 401_499/416_Partition_Equal_Subset_Sum/Solution.rs
impl Solution {
    pub fn can_partition(nums: Vec<i32>) -> bool {
        let mut sum = nums.iter().fold(0, |x1, x2| x1 + x2);
        if sum % 2 == 1 {
            return false;
        }
        sum >>= 1;
        let (n, mut dp) = (nums.len(), vec![false; sum as usize + 1]);
        dp[0] = true;
        for i in 0..n {
            for j in (0..=sum).rev() {
                if j >= nums[i] {
                    let ju = j as usize;
                    dp[ju] = dp[ju] || dp[ju - nums[i] as usize];
                };
            }
        }
        return dp[sum as usize];
    }
}
